File: linearStructure/LinkedList.java
package linearStructure;

class LinkedList {

    int value = -1;
    LinkedList next = null;
    LinkedList prev = null;

    LinkedList(int value) {
        this.value = value;
    }

    LinkedList insert(LinkedList head, int value) {
        LinkedList node = new LinkedList(value);
        if (head == null) {
            head = node;
            return head;
        }
        LinkedList tail = head;
        while (tail.next != null) {
            tail = tail.next;
        }
        tail.next = node;
        node.prev = tail;
        return node;
    }

    LinkedList update(LinkedList head, int value) {
        LinkedList temp = head;
        while (temp != null) {
            if (temp.value == value) {
                return temp;
            } else {
                temp = temp.next;
            }
        }
        print(value + " not found");
        return null;
    }

    LinkedList remove(LinkedList head, int value) {
        LinkedList temp = head;
        while (temp != null) {
            if (temp.value == value) {
                if (temp == head) {
                    if (head.next != null) {
                        head.next.prev = null;
                    }
                    head = head.next;
                } else {
                    temp.prev.next = temp.next;
                }
                break;
            } else {
                temp = temp.next;
            }
        }
        return head;
    }

    public String toString(LinkedList head) {
        String result = "";
        LinkedList temp = head;
        while (temp != null) {
            result += temp.value + " ";
            temp = temp.next;
        }
        return result;
    }

    void print(String msg) {
        System.out.print(msg + "\n");
    }

    LinkedList reverse(LinkedList head) {
        LinkedList temp = head;
        LinkedList preNode = null;
        while (temp != null) {
            LinkedList tempNext = temp.next;
            temp.next = preNode;
            preNode = temp;
            temp = tempNext;
        }
        return preNode;
    }

    public static void main(String[] args) {
        LinkedList head = new LinkedList(8);
        head.insert(head, 10);
        head.print(head.toString(head));
        head.insert(head, 11);
        head.print(head.toString(head));
        head.insert(head, 9);
        head.insert(head, 7);
        head.insert(head, 3);
        head.insert(head, 4);
        head.insert(head, 6);
        head.print(head.toString(head));
        head = head.remove(head, 3);
        head = head.remove(head, 9);
        head.print(head.toString(head));
        head = head.remove(head, 8);
        head.print(head.toString(head));
        head = head.reverse(head);
        head.print(head.toString(head));
    }
}

File: linearStructure/LinkedList.py
from ListNode import ListNode


class LinkedList:
    # 守卫头结点
    def __init__(self):
        self.head = ListNode(-1)
        self.tail = self.head

    def insert(self, value):
        self.tail.next = ListNode(value)
        self.tail = self.tail.next

    def insertFront(self, value):
        temp = self.head.next
        self.head.next = ListNode(value)
        self.head.next.next = temp
        if self.head == self.tail:
            self.tail = self.head.next

    def remove(self, value):
        temp = self.head.next
        prev = self.head
        while temp:
            if temp.val == value:
                prev.next = temp.next
                if temp == self.tail:
                    self.tail = prev
                return temp
            temp = temp.next
            prev = prev.next

        return None
    
    def removeLast(self):
        if self.head.next == None:
            return None
        temp = self.head.next
        prev = self.head
        while temp.next:
            temp = temp.next
            prev = prev.next
        prev.next = None
        return temp

    def find(self, value):
        temp = self.head.next
        while temp:
            if temp.val == value:
                return temp
            temp = temp.next

        return None

    def __str__(self):
        result = ""
        temp = self.head.next
        while temp:
            result += str(temp.val) + "——>"
            temp = temp.next
        result += "None"
        return result


# linkedList = LinkedList()
# print(linkedList)
# linkedList.insertFront(0)
# linkedList.insert(1)
# linkedList.insert(2)
# linkedList.insert(3)
# linkedList.insert(4)
# print(linkedList)
# linkedList.remove(0)
# print(linkedList)
# linkedList.remove(2)
# print(linkedList)
# linkedList.remove(4)
# print(linkedList)
# linkedList.insertFront(0)
# print(linkedList)
# print(linkedList.removeLast().val)
# print(linkedList)
# print(linkedList.find(1).val)
# print(linkedList.find(0).val)
# print(linkedList.find(9))